packages/integration-karma/test/shadow-dom/ShadowRoot.elementsFromPoint/index.spec.js

```diff
@@ -0,0 +1,131 @@
+    // https://crbug.com/1207863#c4
+    let onlyIncludesElementsInImmediateShadowRoot = false;
+    beforeAll(() => {
+        if (process.env.NATIVE_SHADOW) {
+            // detect the Firefox behavior
+            const div = document.createElement('div');
+            div.attachShadow({ mode: 'open' }).innerHTML = '<div>foo</div>';
+            document.body.appendChild(div);
+            const { left, top } = div.shadowRoot.querySelector('div').getBoundingClientRect();
+            const elements = div.shadowRoot.elementsFromPoint(left, top);
+            document.body.removeChild(div);
+            onlyIncludesElementsInImmediateShadowRoot = elements.length === 1;
+        }
+    });
+
+    function test(element, expectedElements) {
+        const { left, top, width, height } = element.getBoundingClientRect();
+        const rootNode = element.getRootNode();
+        const elementsFromPoint = rootNode.elementsFromPoint(left + width / 2, top + height / 2);
+
+        if (onlyIncludesElementsInImmediateShadowRoot) {
+            const mainRootNode = element.getRootNode();
+            expectedElements = expectedElements.filter((el) => el.getRootNode() === mainRootNode);
+        }
+
+        expect(elementsFromPoint).toEqual(expectedElements);
+    }
+
+    // Some browsers have unpredictable behavior for some of these tests
+    const isIE = /(MSIE|Trident)/.test(navigator.userAgent);
+    const isSafari = /Safari/.test(navigator.userAgent) && !/Chrom/.test(navigator.userAgent);
+    const isOldChrome = /Chrom/.test(navigator.userAgent) && process.env.COMPAT;
+    const isOldSafari = isSafari && process.env.COMPAT;
+
+    if (!isOldSafari) {
+        // Safari <v11 doesn't support elementsFromPoint
+        it('non-shadow example', () => {
+            const div = document.createElement('div');
+            div.textContent = 'foo';
+            document.body.appendChild(div);
+
+            const { left, top, width, height } = div.getBoundingClientRect();
+            const elementsFromPoint = document.elementsFromPoint(
+                left + width / 2,
+                top + height / 2
+            );
+
+            expect(elementsFromPoint).toEqual([div, document.body, document.documentElement]);
+        });
+    }
+
+    if (!isIE && !isOldSafari) {
+        // IE returns unpredictable results
+        it('basic shadow example', () => {
+            const elm = createElement('x-container', { is: Container });
+            document.body.appendChild(elm);
+
+            const { body } = document;
+            const html = document.documentElement;
+            const h1 = elm.shadowRoot.querySelector('h1');
+            const inContainer = elm.shadowRoot.querySelector('.in-container');
+            const slottable = elm.shadowRoot.querySelector('x-slottable');
+            const slotted = elm.shadowRoot.querySelector('.slotted');
+
+            test(elm, [elm, body, html]);
```

Comment:
I would like to better understand how this test works in practice. For example `elm` the parent element for `inContainer`, `slottable` and `slotted`.  How can you ensure that when you invoke `elementsFromPoint` at the center of the middle of the `elm` you don't inadvertently hit one of the children's elements?

I would have expected the fixture components to have a bunch of CSS with absolute positioning to avoid this issue.

Reply:
You're right, this is probably why the test is inconsistent across browsers. I'll fix it to use more consistent CSS, which may hopefully fix the cross-browser issues.

Reply:
Yep, that's the reason for the inconsistency. 🤦  No more weird browser hacks needed. I should have remembered that browsers can never agree on pixel-perfect text layout stuff.